File: ios/KidCanvasApp/Views/ReactionsAndComments.swift
import SwiftUI

/// A row of the five allowed reactions, showing counts and what this user has
/// already left. Taps are optimistic — the count moves immediately.
struct ReactionBar: View {
    let artworkId: UUID
    let service: ArtworkService

    @State private var counts: [String: Int] = [:]
    @State private var mine: Set<String> = []

    var body: some View {
        HStack(spacing: 8) {
            ForEach(Reaction.allCases) { reaction in
                ReactionButton(
                    reaction: reaction,
                    count: counts[reaction.rawValue] ?? 0,
                    isMine: mine.contains(reaction.rawValue)
                ) {
                    toggle(reaction)
                }
            }
        }
        .task(id: artworkId) { await load() }
    }

    private func load() async {
        async let countsResult = try? service.reactionCounts(artworkId: artworkId)
        async let mineResult = try? service.myReactions(artworkId: artworkId)
        let (fetchedCounts, fetchedMine) = await (countsResult, mineResult)
        if let fetchedCounts {
            counts = Dictionary(uniqueKeysWithValues: fetchedCounts.map { ($0.emojiType, $0.count) })
        }
        if let fetchedMine { mine = fetchedMine }
    }

    private func toggle(_ reaction: Reaction) {
        let key = reaction.rawValue
        let wasMine = mine.contains(key)

        // Optimistic: reflect the tap now, reconcile after the round trip.
        if wasMine {
            mine.remove(key)
            counts[key] = max(0, (counts[key] ?? 1) - 1)
        } else {
            mine.insert(key)
            counts[key] = (counts[key] ?? 0) + 1
        }

        Task {
            do {
                if wasMine {
                    try await service.removeReaction(reaction, artworkId: artworkId)
                } else {
                    try await service.addReaction(reaction, artworkId: artworkId)
                }
            } catch {
                await load()
            }
        }
    }
}

struct ReactionButton: View {
    let reaction: Reaction
    let count: Int
    let isMine: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: 4) {
                Text(reaction.rawValue)
                    .font(.callout)
                if count > 0 {
                    Text("\(count)")
                        .font(.caption.weight(.semibold))
                        .foregroundColor(isMine ? .white : .secondary)
                }
            }
            .padding(.horizontal, 10)
            .padding(.vertical, 7)
            .background(
                isMine
                    ? AnyShapeStyle(LinearGradient(colors: [.pink, .purple],
                                                   startPoint: .leading, endPoint: .trailing))
                    : AnyShapeStyle(Color.subtleFill),
                in: Capsule()
            )
        }
        .buttonStyle(.plain)
        .sensoryFeedback(.impact(weight: .light), trigger: isMine)
        .accessibilityLabel("\(reaction.label)\(count > 0 ? ", \(count)" : "")")
        .accessibilityHint(isMine ? "Tap to remove your reaction" : "Tap to react")
    }
}

struct CommentsSection: View {
    let artworkId: UUID
    let familyId: UUID
    let service: ArtworkService

    @State private var comments: [ArtworkComment] = []
    @State private var draft = ""
    @State private var isSending = false
    @State private var role: String?
    @State private var deleteError: String?

    /// Authors can always remove their own comment. Owners and parents can remove
    /// anyone's, which is what makes this moderation rather than just an undo --
    /// App Store Guideline 1.2 asks for a way to take content down.
    private func canDelete(_ comment: ArtworkComment) -> Bool {
        comment.userId == service.currentUserID || role == "owner" || role == "parent"
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            Label(comments.isEmpty ? "Comments" : "Comments (\(comments.count))",
                  systemImage: "bubble.left.and.bubble.right")
                .font(.subheadline.bold())
                .foregroundColor(.secondary)

            ForEach(comments) { comment in
                CommentRow(text: comment.text, date: comment.createdAt)
                    .contextMenu {
                        if canDelete(comment) {
                            Button(role: .destructive) {
                                delete(comment)
                            } label: {
                                Label("Delete Comment", systemImage: "trash")
                            }
                        }
                    }
            }

            HStack(spacing: 8) {
                TextField("Add a comment", text: $draft, axis: .vertical)
                    .lineLimit(1...3)
                    .padding(10)
                    .background(Color.subtleFill, in: RoundedRectangle(cornerRadius: 10))
                Button {
                    send()
                } label: {
                    Image(systemName: isSending ? "hourglass" : "arrow.up.circle.fill")
                        .font(.title2)
                        .foregroundStyle(canSend ? Color.pink : Color.secondary)
                }
                .disabled(!canSend)
                .accessibilityLabel("Post Comment")
            }
        }
        .task(id: artworkId) {
            await load()
            role = await service.familyRole(familyId: familyId)
        }
        .alert("Couldn't delete that comment",
               isPresented: .constant(deleteError != nil),
               actions: { Button("OK") { deleteError = nil } },
               message: { Text(deleteError ?? "") })
    }

    private func delete(_ comment: ArtworkComment) {
        Task {
            do {
                try await service.deleteComment(id: comment.id)
                await load()
            } catch {
                // Surface it rather than leaving the comment silently in place.
                deleteError = error.localizedDescription
            }
        }
    }

    private var canSend: Bool {
        !draft.trimmed.isEmpty && !isSending
    }

    private func load() async {
        comments = (try? await service.comments(artworkId: artworkId)) ?? []
    }

    private func send() {
        let text = draft.trimmed
        guard !text.isEmpty else { return }
        isSending = true
        Task {
            try? await service.addComment(text, artworkId: artworkId)
            draft = ""
            await load()
            isSending = false
        }
    }
}

struct CommentRow: View {
    let text: String
    let date: Date

    var body: some View {
        VStack(alignment: .leading, spacing: 2) {
            Text(text)
                .font(.subheadline)
                .fixedSize(horizontal: false, vertical: true)
            Text(date, format: .relative(presentation: .named))
                .font(.caption2)
                .foregroundColor(.secondary)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(12)
        .background(Color.faintFill, in: RoundedRectangle(cornerRadius: 12))
    }
}

File: android/app/src/main/java/app/kidcanvas/ui/social/ReactionsAndComments.kt
package app.kidcanvas.ui.social

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.widget.Toast
import androidx.compose.ui.platform.LocalContext

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.combinedClickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.draw.clip
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.Send
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.hapticfeedback.HapticFeedbackType
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import app.kidcanvas.data.ArtworkComment
import app.kidcanvas.data.COMMENT_MAX_CHARS
import app.kidcanvas.data.KidCanvasRepository
import app.kidcanvas.data.Reaction
import app.kidcanvas.data.Roles
import app.kidcanvas.ui.common.BrandGradient
import app.kidcanvas.ui.common.Formatting
import app.kidcanvas.ui.common.friendlyError
import kotlinx.coroutines.async
import kotlinx.coroutines.launch

/**
 * The five allowed reactions with counts and what this user already left.
 * Taps are optimistic like iOS: the count moves now, and a failed round trip
 * reloads the real numbers instead of leaving a lie on screen.
 */
@Composable
fun ReactionBar(artworkId: String, repo: KidCanvasRepository, modifier: Modifier = Modifier) {
    var counts by remember(artworkId) { mutableStateOf<Map<String, Long>>(emptyMap()) }
    var mine by remember(artworkId) { mutableStateOf<Set<String>>(emptySet()) }
    val scope = rememberCoroutineScope()
    val haptics = LocalHapticFeedback.current

    suspend fun load() {
        val c = scope.async { runCatching { repo.reactionCounts(artworkId) }.getOrNull() }
        val m = scope.async { runCatching { repo.myReactions(artworkId) }.getOrNull() }
        c.await()?.let { counts = it }
        m.await()?.let { mine = it }
    }

    LaunchedEffect(artworkId) { load() }

    // Scrolls rather than wraps: five chips with counts overflow a 320dp phone
    // at large font scale, and a second row would push comments down the card.
    Row(modifier.horizontalScroll(rememberScrollState()), horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        Reaction.entries.forEach { reaction ->
            val key = reaction.emoji
            val count = counts[key] ?: 0
            val isMine = key in mine
            Row(
                verticalAlignment = Alignment.CenterVertically,
                modifier = Modifier
                    .clip(RoundedCornerShape(50))
                    .then(
                        if (isMine) Modifier.background(BrandGradient, RoundedCornerShape(50))
                        else Modifier.background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.07f), RoundedCornerShape(50))
                    )
                    .clickable {
                        haptics.performHapticFeedback(HapticFeedbackType.TextHandleMove)
                        val wasMine = isMine
                        if (wasMine) {
                            mine = mine - key
                            counts = counts + (key to maxOf(0, (counts[key] ?: 1) - 1))
                        } else {
                            mine = mine + key
                            counts = counts + (key to (counts[key] ?: 0) + 1)
                        }
                        scope.launch {
                            try {
                                if (wasMine) repo.removeReaction(reaction, artworkId) else repo.addReaction(reaction, artworkId)
                            } catch (_: Throwable) {
                                load()
                            }
                        }
                    }
                    .semantics {
                        contentDescription = reaction.label + (if (count > 0) ", $count" else "") +
                            if (isMine) ". Tap to remove your reaction" else ". Tap to react"
                    }
                    .padding(horizontal = 10.dp, vertical = 7.dp),
            ) {
                Text(reaction.emoji, style = MaterialTheme.typography.bodyLarge)
                if (count > 0) {
                    Spacer(Modifier.width(4.dp))
                    Text(
                        "$count",
                        style = MaterialTheme.typography.labelMedium,
                        fontWeight = FontWeight.SemiBold,
                        color = if (isMine) Color.White else MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }
        }
    }
}

/**
 * Comments in the order they were written, with a box to add one. Authors can
 * delete their own; owners and parents can delete anyone's, which is the
 * moderation path both app stores expect from user-generated content.
 */
@OptIn(ExperimentalFoundationApi::class)
@Composable
fun CommentsSection(artworkId: String, role: String?, repo: KidCanvasRepository) {
    var comments by remember(artworkId) { mutableStateOf<List<ArtworkComment>>(emptyList()) }
    var draft by rememberSaveable(artworkId) { mutableStateOf("") }
    var sending by remember { mutableStateOf(false) }
    var pendingAction by remember { mutableStateOf<ArtworkComment?>(null) }
    var pendingDelete by remember { mutableStateOf<ArtworkComment?>(null) }
    var errorText by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    val context = LocalContext.current
    val me = repo.currentUserId()

    suspend fun load() {
        runCatching { repo.comments(artworkId) }.onSuccess { comments = it }
    }
    LaunchedEffect(artworkId) { load() }

    fun canDelete(c: ArtworkComment) = c.userId == me || Roles.canManage(role)

    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        Text(
            if (comments.isEmpty()) "Comments" else "Comments (${comments.size})",
            style = MaterialTheme.typography.titleSmall,
            fontWeight = FontWeight.Bold,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )

        comments.forEach { comment ->
            Column(
                Modifier
                    .fillMaxWidth()
                    .clip(RoundedCornerShape(12.dp))
                    .background(MaterialTheme.colorScheme.onSurface.copy(alpha = 0.04f))
                    .combinedClickable(
                        onClick = {},
                        onLongClick = { pendingAction = comment },
                        onLongClickLabel = "Comment options",
                    )
                    .padding(12.dp),
            ) {
                Text(comment.text, style = MaterialTheme.typography.bodyMedium)
                Spacer(Modifier.height(2.dp))
                Text(
                    Formatting.relative(comment.createdAt),
                    style = MaterialTheme.typography.labelSmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
            }
        }

        val canSend = draft.isNotBlank() && !sending
        Row(verticalAlignment = Alignment.CenterVertically) {
            OutlinedTextField(
                value = draft,
                // Capped as typed so the 500 character CHECK constraint can never
                // reject a comment someone spent a while writing.
                onValueChange = { draft = it.take(COMMENT_MAX_CHARS) },
                placeholder = { Text("Add a comment") },
                maxLines = 3,
                supportingText = if (draft.length > COMMENT_MAX_CHARS - 50) {
                    { Text("${draft.length} / $COMMENT_MAX_CHARS") }
                } else null,
                modifier = Modifier.weight(1f),
            )
            IconButton(
                enabled = canSend,
                onClick = {
                    val text = draft.trim()
                    sending = true
                    scope.launch {
                        try {
                            repo.addComment(text, artworkId)
                            draft = ""
                            load()
                        } catch (t: Throwable) {
                            errorText = friendlyError(t, "Could not post that comment. Try again.")
                        } finally {
                            sending = false
                        }
                    }
                },
            ) {
                if (sending) CircularProgressIndicator(Modifier.size(20.dp), strokeWidth = 2.dp)
                else Icon(Icons.AutoMirrored.Filled.Send, contentDescription = "Post comment",
                    tint = if (canSend) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.onSurfaceVariant)
            }
        }
        if (comments.isNotEmpty()) {
            Text(
                if (comments.any { canDelete(it) }) "Press and hold a comment to delete or report it."
                else "Press and hold a comment to report it.",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
    }

    pendingAction?.let { comment ->
        AlertDialog(
            onDismissRequest = { pendingAction = null },
            title = { Text("Comment") },
            text = { Text(comment.text, maxLines = 4) },
            confirmButton = {
                Row {
                    if (comment.userId != me) {
                        TextButton(onClick = {
                            pendingAction = null
                            reportComment(context, comment)
                        }) { Text("Report") }
                    }
                    if (canDelete(comment)) {
                        TextButton(onClick = { pendingAction = null; pendingDelete = comment }) {
                            Text("Delete", color = MaterialTheme.colorScheme.error)
                        }
                    }
                }
            },
            dismissButton = { TextButton(onClick = { pendingAction = null }) { Text("Cancel") } },
        )
    }

    pendingDelete?.let { comment ->
        AlertDialog(
            onDismissRequest = { pendingDelete = null },
            title = { Text("Delete this comment?") },
            text = { Text("It will be removed for everyone in the family.") },
            confirmButton = {
                TextButton(onClick = {
                    pendingDelete = null
                    scope.launch {
                        try {
                            if (!repo.deleteComment(comment.id)) errorText = "Only the author, owners and parents can delete comments."
                            load()
                        } catch (t: Throwable) {
                            // Surface it rather than leaving the comment silently in place.
                            errorText = friendlyError(t, "Could not delete that comment.")
                        }
                    }
                }) { Text("Delete", color = MaterialTheme.colorScheme.error) }
            },
            dismissButton = { TextButton(onClick = { pendingDelete = null }) { Text("Cancel") } },
        )
    }

    errorText?.let {
        AlertDialog(
            onDismissRequest = { errorText = null },
            text = { Text(it) },
            confirmButton = { TextButton(onClick = { errorText = null }) { Text("OK") } },
        )
    }
}

/**
 * Play's user-generated content policy asks for an in-app way to report
 * content. There is no reports table in the schema, so this opens an email to
 * support with the ids filled in; the person sees and sends it themselves.
 */
private fun reportComment(context: Context, comment: ArtworkComment) {
    val body = "I want to report a comment in KidCanvas.\n\n" +
        "Comment id: ${comment.id}\nArtwork id: ${comment.artworkId}\n\nWhat is wrong with it:\n"
    val intent = Intent(Intent.ACTION_SENDTO, Uri.parse("mailto:")).apply {
        putExtra(Intent.EXTRA_EMAIL, arrayOf(SUPPORT_EMAIL))
        putExtra(Intent.EXTRA_SUBJECT, "Report a comment")
        putExtra(Intent.EXTRA_TEXT, body)
    }
    runCatching { context.startActivity(intent) }.onFailure {
        Toast.makeText(context, "No email app found. Write to $SUPPORT_EMAIL.", Toast.LENGTH_LONG).show()
    }
}

private const val SUPPORT_EMAIL = "support@kidcanvas.app"

